ludwig/features/base_feature.py

```diff
@@ -134,11 +134,11 @@ def train_loss(self, targets, predictions):
         return self.train_loss_function(targets, predictions)
 
     def eval_loss(self, targets, predictions):
-        return self.eval_loss_function(targets, predictions)
+        return self.eval_loss_function(targets, predictions['predictions'])
```

Comment:
Here and in `update_metrics` I would rather pass directly `predictions['predictions']` when those functions are called after than having them get the `'predictions'` key. It's a minor thing, but in theory this make these functions reusables for any vector you trow at them independently, while doing this way they make an assumption that predictions is not just a vector but a dictionary that contains a specific key.

Reply:
I believe this is resolved.  However, you should probably pay close attention to how I resolved it.  This issue came about with the implementation of `ECE.predict()` and how predictions are now presented as a dictionary values,e.g.,
```
{'my_numerical_feature': { 'predictions': [...], 'logits': [...] }

{'my_binary_feature': {'predictions': [...], 'probability': [...], 'logits': [...]}

{'my_categorical_feature': {'predictions': [...], 'probability': [...],n'probabilities': [...], 'logits': [...]}}
```

From what I can tell depending on the type output feature and specific metric, different prediction values (e.g., logits, predictions) are used.  To accommodate this, I decided to override the `OutputFeature.update_metrics()` method with the `<Type>OutputFeature.update_metrics()` specific method.  

If this is a reasonable approach, then it may make sense to change `OutputFeature.update_metrics()` to be an abstract method and let the specific `<Type>OutputFeature` to provide an implementation.

I'll leave this unresolved until you have a chance to look at it.

Reply:
It makes sense, but you are probably right at using the dereference of the dictionary inside the loss function, as the loss function is defined  within the `<Type>OutputFeature` class so the developer of that class can use different elements of the prediction dictionary depending on the different type of loss. So, my bad, I guess we can keep you previous solution. Sorry for pushing you into a wrong direction.

Reply:
NP  I've been through enough of this kind of work to know that sometimes we head down the wrong path and have to adjust.  Flexibility is the key to happiness. :-)

Anyway, I'd like to confirm current guidance.  When you write:
> It makes sense, ...

Are you referring to the suggestion 
> From what I can tell depending on the type output feature and specific metric, different prediction values (e.g., logits, predictions) are used. To accommodate this, I decided to override the OutputFeature.update_metrics() method with the \<Type\>OutputFeature.update_metrics() specific method.
> 
> If this is a reasonable approach, then it may make sense to change OutputFeature.update_metrics() to be an abstract method and let the specific \<Type\>OutputFeature to provide an implementation.

If this is correct, then I'll implement the change to `OutputFeature.update_metrics()` to be an abstract method.  OTOH, if I'm mistaken, I need more guidance.

Re: 
>  dereference of the dictionary inside the loss function, as the loss function is defined within the <Type>OutputFeature class 

I'll go back to the prior method where "deference of the dictionary is inside the loss function".